slurmutils/models/model.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2024 Canonical Ltd.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Base classes and methods for composing Slurm data models."""
+
+__all__ = ["BaseModel", "LineInterface", "format_key", "generate_descriptors"]
+
+import copy
+import json
+import re
+from abc import ABC, abstractmethod
+from typing import Any, Dict
+
+from ..exceptions import ModelError
+
+_acronym = re.compile(r"(?<=[A-Z])(?=[A-Z][a-z])")
+_camelize = re.compile(r"(?<=[a-z0-9])(?=[A-Z])")
+
+
+def format_key(key: str) -> str:
+    """Format Slurm configuration keys from SlurmCASe to camelCase.
+
+    Args:
+        key: Configuration key to format into camel case.
+
+    Notes:
+       Slurm configuration syntax does not follow proper PascalCasing
+       format, so we cannot put keys directly through a kebab case converter
+       to get the desired format. Some additional processing is needed for
+       certain keys before the key can properly camelized.
+
+       For example, without additional preprocessing, the key `CPUs` will
+       become `cp-us` if put through a caramelize with being preformatted to `Cpus`.
+    """
+    if "CPUs" in key:
+        key = key.replace("CPUs", "Cpus")
+    key = _acronym.sub(r"_", key)
+    return _camelize.sub(r"_", key).lower()
+
+
+def generate_descriptors(opt: str):
```

Comment:
Type annotation?

Reply:
Yeah - I can add one, just returns tuple of callables.

Reply:
Added type annotation in https://github.com/charmed-hpc/slurmutils/pull/15/commits/67a1bdc9be8fb1b5ea42cba586883cdb6465a183. Further work will need to be done in a separate PR to enable project wide type checking. I did initially set up pyright, but it complained quite a bit and the dynamically defined descriptors on the data models :sweat_smile: